### ISSUE TITLE
y-boundaries in 3d Laplace solvers not set if not using parallel slices

### ISSUE THREAD

**Issue by johnomotani** [boutproject/BOUT-dev#2253]
If a 3d Laplace solver (`LaplacePetsc3dAmg`, or `LaplaceHypre3d` once that gets merged) is used in a simulation that does not calculate parallel slices, it will not apply any parallel boundary condition to its result. At the moment it is up to the user to deal with this, and there is not even a warning printed. Possibly we should at least add a warning in the constructor? I'd rather not make it an exception, because I'd (eventually) like to use these solvers in STORM, which will probably have to just deal with the boundary conditions specially since it uses `toFieldAligned()`/`fromFieldAligned()` instead of parallel slices.

**Comment by ZedThree** [comment]
If it doesn't have parallel slices, should it not just set the boundaries in the non-slice cells, i.e. don't use `yup/ydown`?

**Comment by johnomotani** [comment]
> If it doesn't have parallel slices, should it not just set the boundaries in the non-slice cells, i.e. don't use `yup/ydown`?

Ideally yes, but the boundary value calculated is at yup/ydown positions, so needs a shift in z that as far as I know there's not a convenient way to do at the moment... Also, if not using parallel slices, setting the y boundaries in the non-slice cells is slightly non-optimal. At least for the STORM case, we'll take the result `phi`, create a `phi_aligned = toFieldAligned(phi);`, and the parallel boundary condition is only needed on `phi_aligned`, so it's probably most efficient to either re-apply the boundary condition to `phi_aligned`, or shift directly from the boundary cells returned by the Laplace solver to phi_aligned (otherwise the boundary cells get shifted twice: first to non-aligned, then to field-aligned).

Basically I think we need to get these solvers into use and see what's needed once we have some experience running them. The issue is more of a reminder to sort this out at that point (sorry, in hindsight I should have said this in the original issue).

Could also consider what's best to do if using `ParallelTransformIdentity`, but I guess it's not likely anyone will have a use for a 3d solver when using `ParallelTransformIdentity`, so seems like a low priority.
